Memory violation in coap

`make -C test test_asan TEST_FILTER=coap`

```
CC	unit_test_asan
RUN	unit_test_asan coap
=================================================================
==15382==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7fff5ab81fc3 at pc 0x000105144d93 bp 0x7fff5ab80c70 sp 0x7fff5ab80c68
READ of size 1 at 0x7fff5ab81fc3 thread T0
    #0 0x105144d92 in coap_get_options /Users/mkm/Projects/cesanta/fossa/test/src/coap.c:246:3
    #1 0x105142b54 in ns_coap_parse /Users/mkm/Projects/cesanta/fossa/test/src/coap.c:348:14
    #2 0x1050a79e9 in test_coap /Users/mkm/Projects/cesanta/fossa/test/unit_test.c:2212:9
    #3 0x105083348 in run_tests /Users/mkm/Projects/cesanta/fossa/test/unit_test.c:2487:3
    #4 0x10507df16 in main /Users/mkm/Projects/cesanta/fossa/test/unit_test.c:2497:14
    #5 0x7fff8eda65fc in start (/usr/lib/system/libdyld.dylib+0x35fc)
    #6 0x1 (<unknown module>)

Address 0x7fff5ab81fc3 is located in stack of thread T0 at offset 323 in frame
    #0 0x1050a549f in test_coap /Users/mkm/Projects/cesanta/fossa/test/unit_test.c:2115

  This frame has 24 object(s):
    [32, 40) 'retval'
    [64, 88) 'packet_in'
    [128, 152) 'packet_out'
    [192, 256) 'cm'
    [288, 292) 'res'
    [304, 323) 'coap_packet_1' <== Memory access at offset 323 overflows this variable
    [368, 372) 'coap_packet_2'
    [384, 513) 'coap_packet_3'
    [592, 596) 'coap_packet_4'
    [608, 638) 'coap_packet_5'
    [672, 676) 'coap_packet_6'
    [688, 718) 'coap_packet_7'
    [752, 756) 'coap_packet_2_broken'
    [768, 772) 'coap_packet_2_broken1098'
    [784, 786) 'value16'
    [800, 832) 'mgr'
    [864, 872) 'nc'
    [896, 904) 'address'
    [928, 936) 'res1488'
    [960, 992) 'mgr1489'
    [1024, 1032) 'nc1'
    [1056, 1064) 'nc2'
    [1088, 1096) 'address1490'
    [1120, 1124) 'i'
HINT: this may be a false positive if your program uses some custom stack unwind mechanism or swapcontext
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow /Users/mkm/Projects/cesanta/fossa/test/src/coap.c:246 coap_get_options
Shadow bytes around the buggy address:
  0x1fffeb5703a0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1fffeb5703b0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1fffeb5703c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1fffeb5703d0: f1 f1 f1 f1 00 f2 f2 f2 00 00 00 f2 f2 f2 f2 f2
  0x1fffeb5703e0: 00 00 00 f2 f2 f2 f2 f2 00 00 00 00 00 00 00 00
=>0x1fffeb5703f0: f2 f2 f2 f2 04 f2 00 00[03]f2 f2 f2 f2 f2 04 f2
  0x1fffeb570400: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1fffeb570410: 01 f2 f2 f2 f2 f2 f2 f2 f2 f2 04 f2 00 00 00 06
  0x1fffeb570420: f2 f2 f2 f2 04 f2 00 00 00 06 f2 f2 f2 f2 04 f2
  0x1fffeb570430: 04 f2 02 f2 00 00 00 00 f2 f2 f2 f2 00 f2 f2 f2
  0x1fffeb570440: 00 f2 f2 f2 00 f2 f2 f2 00 00 00 00 f2 f2 f2 f2
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  ASan internal:           fe
==15382==ABORTING
make: *** [test_asan] Error 1
```

--- a/fossa.c
+++ b/fossa.c
@@ -5172,7 +5172,7 @@ void MD5_Final(unsigned char digest[16], MD5_CTX *ctx) {
 
 #ifdef NS_ENABLE_COAP
 
-/* 
+/*
  * CoAP implementation.
  *
  * General CoAP message format:
@@ -5186,7 +5186,7 @@ void MD5_Final(unsigned char digest[16], MD5_CTX *ctx) {
 /* TODO(alashkin): remove this #include after dev completion */
 #include <stdlib.h>
 
-/* 
+/*
  * Frees the memory allocated for options,
  * if cm paramater doesn't contain any option does nothing.
  */
@@ -5198,7 +5198,7 @@ void ns_coap_free_options(struct ns_coap_message *cm) {
   }
 }
 
-/* 
+/*
  * Adds new option to ns_coap_message structure.
  * Returns pointer to newly created options.
  */
@@ -5215,7 +5215,7 @@ struct ns_coap_option *ns_coap_add_option(struct ns_coap_message *cm,
   if (cm->options == NULL) {
     cm->options = cm->options_tail = new_option;
   } else {
-    /* 
+    /*
      * A very simple attention to help clients to compose options:
      * CoAP wants to see options ASC ordered.
      * Could be change by using sort in coap_compose
@@ -5250,8 +5250,8 @@ struct ns_coap_option *ns_coap_add_option(struct ns_coap_message *cm,
   return new_option;
 }
 
-/* 
- * Fills CoAP header in ns_coap_message. 
+/*
+ * Fills CoAP header in ns_coap_message.
  *
  * Helper function.
  */
@@ -5335,8 +5335,8 @@ static char *coap_get_token(char *ptr, struct iobuf *io,
   return ptr;
 }
 
-/* 
- * Returns Option Delta or Length. 
+/*
+ * Returns Option Delta or Length.
  *
  * Helper function.
  */
@@ -5370,7 +5370,7 @@ static int coap_get_ext_opt(char* ptr, struct iobuf *io, uint16_t* opt_info) {
   return ret;
 }
 
-/* 
+/*
  * Fills options in ns_coap_message.
  *
  * Helper function.
@@ -5396,7 +5396,7 @@ static char *coap_get_options(char* ptr, struct iobuf *io,
   }
 
   /* 0xFF is payload marker */
-  while ((uint8_t)*ptr != 0xFF && ptr < io->buf + io->len) {
+  while (ptr < io->buf + io->len && (uint8_t)*ptr != 0xFF) {
     uint16_t option_delta, option_lenght;
     int optinfo_len;
 
@@ -5512,8 +5512,8 @@ uint32_t ns_coap_parse(struct iobuf *io, struct ns_coap_message *cm) {
   return cm->flags;
 }
 
-/* 
- * Calculates extended size of given Opt Number/Length in coap message. 
+/*
+ * Calculates extended size of given Opt Number/Length in coap message.
  *
  * Helper function.
  */
@@ -5529,8 +5529,8 @@ static size_t coap_get_ext_opt_size(uint32_t value) {
   return ret;
 }
 
-/* 
- * Splits given Opt Number/Length into base and ext values. 
+/*
+ * Splits given Opt Number/Length into base and ext values.
  *
  * Helper function.
  */
@@ -5553,8 +5553,8 @@ static int coap_split_opt(uint32_t value, uint8_t *base, uint16_t *ext) {
   return ret;
 }
 
-/* 
- * Puts uint16_t (in network order) into given char stream. 
+/*
+ * Puts uint16_t (in network order) into given char stream.
  *
  * Helper function.
  */
@@ -5566,8 +5566,8 @@ static char *coap_add_uint16(char *ptr, uint16_t val) {
   return ptr;
 }
 
-/* 
- * Puts extended value of Opt Number/Length into given char stream. 
+/*
+ * Puts extended value of Opt Number/Length into given char stream.
  *
  * Helper function.
  */
@@ -5582,8 +5582,8 @@ static char *coap_add_opt_info(char *ptr, uint16_t val, size_t len) {
   return ptr;
 }
 
-/* 
- * Verifies given ns_coap_message and calculates message size for it. 
+/*
+ * Verifies given ns_coap_message and calculates message size for it.
  *
  * Helper function.
  */
@@ -5617,7 +5617,7 @@ static uint32_t coap_calculate_packet_size(struct ns_coap_message *cm,
     *len += 1;  /* basic delta/length */
     *len += coap_get_ext_opt_size(opt->number);
     *len += coap_get_ext_opt_size((uint32_t)opt->value.len);
-    /* 
+    /*
      * Current implementation performs check if
      * option_number > previous option_number and produces an error
      * TODO(alashkin): write design doc with limitations
@@ -5713,7 +5713,7 @@ uint32_t ns_coap_compose(struct ns_coap_message *cm, struct iobuf *io) {
   return 0;
 }
 
-/* 
+/*
  * Composes CoAP message from ns_coap_message
  * and sends it into 'nc' connection.
  */

--- a/src/coap.c
+++ b/src/coap.c
@@ -17,7 +17,7 @@
 
 #ifdef NS_ENABLE_COAP
 
-/* 
+/*
  * CoAP implementation.
  *
  * General CoAP message format:
@@ -34,7 +34,7 @@
 #include "iobuf.h"
 #include "coap.h"
 
-/* 
+/*
  * Frees the memory allocated for options,
  * if cm paramater doesn't contain any option does nothing.
  */
@@ -46,7 +46,7 @@ void ns_coap_free_options(struct ns_coap_message *cm) {
   }
 }
 
-/* 
+/*
  * Adds new option to ns_coap_message structure.
  * Returns pointer to newly created options.
  */
@@ -63,7 +63,7 @@ struct ns_coap_option *ns_coap_add_option(struct ns_coap_message *cm,
   if (cm->options == NULL) {
     cm->options = cm->options_tail = new_option;
   } else {
-    /* 
+    /*
      * A very simple attention to help clients to compose options:
      * CoAP wants to see options ASC ordered.
      * Could be change by using sort in coap_compose
@@ -98,8 +98,8 @@ struct ns_coap_option *ns_coap_add_option(struct ns_coap_message *cm,
   return new_option;
 }
 
-/* 
- * Fills CoAP header in ns_coap_message. 
+/*
+ * Fills CoAP header in ns_coap_message.
  *
  * Helper function.
  */
@@ -183,8 +183,8 @@ static char *coap_get_token(char *ptr, struct iobuf *io,
   return ptr;
 }
 
-/* 
- * Returns Option Delta or Length. 
+/*
+ * Returns Option Delta or Length.
  *
  * Helper function.
  */
@@ -218,7 +218,7 @@ static int coap_get_ext_opt(char* ptr, struct iobuf *io, uint16_t* opt_info) {
   return ret;
 }
 
-/* 
+/*
  * Fills options in ns_coap_message.
  *
  * Helper function.
@@ -244,7 +244,7 @@ static char *coap_get_options(char* ptr, struct iobuf *io,
   }
 
   /* 0xFF is payload marker */
-  while ((uint8_t)*ptr != 0xFF && ptr < io->buf + io->len) {
+  while (ptr < io->buf + io->len && (uint8_t)*ptr != 0xFF) {
     uint16_t option_delta, option_lenght;
     int optinfo_len;
 
@@ -360,8 +360,8 @@ uint32_t ns_coap_parse(struct iobuf *io, struct ns_coap_message *cm) {
   return cm->flags;
 }
 
-/* 
- * Calculates extended size of given Opt Number/Length in coap message. 
+/*
+ * Calculates extended size of given Opt Number/Length in coap message.
  *
  * Helper function.
  */
@@ -377,8 +377,8 @@ static size_t coap_get_ext_opt_size(uint32_t value) {
   return ret;
 }
 
-/* 
- * Splits given Opt Number/Length into base and ext values. 
+/*
+ * Splits given Opt Number/Length into base and ext values.
  *
  * Helper function.
  */
@@ -401,8 +401,8 @@ static int coap_split_opt(uint32_t value, uint8_t *base, uint16_t *ext) {
   return ret;
 }
 
-/* 
- * Puts uint16_t (in network order) into given char stream. 
+/*
+ * Puts uint16_t (in network order) into given char stream.
  *
  * Helper function.
  */
@@ -414,8 +414,8 @@ static char *coap_add_uint16(char *ptr, uint16_t val) {
   return ptr;
 }
 
-/* 
- * Puts extended value of Opt Number/Length into given char stream. 
+/*
+ * Puts extended value of Opt Number/Length into given char stream.
  *
  * Helper function.
  */
@@ -430,8 +430,8 @@ static char *coap_add_opt_info(char *ptr, uint16_t val, size_t len) {
   return ptr;
 }
 
-/* 
- * Verifies given ns_coap_message and calculates message size for it. 
+/*
+ * Verifies given ns_coap_message and calculates message size for it.
  *
  * Helper function.
  */
@@ -465,7 +465,7 @@ static uint32_t coap_calculate_packet_size(struct ns_coap_message *cm,
     *len += 1;  /* basic delta/length */
     *len += coap_get_ext_opt_size(opt->number);
     *len += coap_get_ext_opt_size((uint32_t)opt->value.len);
-    /* 
+    /*
      * Current implementation performs check if
      * option_number > previous option_number and produces an error
      * TODO(alashkin): write design doc with limitations
@@ -561,7 +561,7 @@ uint32_t ns_coap_compose(struct ns_coap_message *cm, struct iobuf *io) {
   return 0;
 }
 
-/* 
+/*
  * Composes CoAP message from ns_coap_message
  * and sends it into 'nc' connection.
  */

--- a/test/unit_test.c
+++ b/test/unit_test.c
@@ -2394,11 +2394,12 @@ static const char *test_coap(void) {
   {
     struct ns_mgr mgr;
     struct ns_connection *nc;
-    const char *address = "tcp://127.0.0.1:8080";
+    const char *address = "tcp://127.0.0.1:8686";
 
     ns_mgr_init(&mgr, 0);
 
     nc = ns_bind(&mgr, address, coap_handler_1);
+    ASSERT(nc != NULL);
     ASSERT(ns_set_protocol_coap(nc) == -1);
 
     ns_mgr_free(&mgr);

